usr.bin/touch/touch.1
.\"	$NetBSD: touch.1,v 1.16 2003/08/07 11:16:44 agc Exp $
.\"
.\" Copyright (c) 1991, 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" This code is derived from software contributed to Berkeley by
.\" the Institute of Electrical and Electronics Engineers, Inc.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"     @(#)touch.1	8.3 (Berkeley) 4/28/95
.\"
.Dd April 28, 1995
.Dt TOUCH 1
.Os
.Sh NAME
.Nm touch
.Nd change file access and modification times
.Sh SYNOPSIS
.Nm
.Op Fl acfhim
.Op Fl r Ar file
.Op Fl t Ar [[CC]YY]MMDDhhmm[.SS]
.Op Fl d Ar inc
.Ar file ...
.Sh DESCRIPTION
The
.Nm
utility sets the modification and access times of files to the
current time of day.
If the file doesn't exist, it is created with default permissions.
.Pp
The following options are available:
.Bl -tag -width Ds
.It Fl a
Change the access time of the file.
The modification time of the file is not changed unless the
.Fl m
flag is also specified.
.It Fl c
Do not create the file if it does not exist.
The
.Nm
utility does not treat this as an error.
No error messages are displayed and the exit value is not affected.
.It Fl d Ar inc
Specify
.Ar inc ,
a time increment
.\" This should be just
.\"	.Pf ( Sq d Ns
.\"	elta) in seconds...
.\" but that doesn't work with 1.4T's -mandoc.  (It does work with 4.0.1's.)
.Sm off
.Xo
.Pf ( Sq d
.No elta)
.Xc
.Sm on
in seconds, which is added to the time used for each file to give
the time used for the next file; other options which control the
timestamp to be used specify the time for the first-affected file.
This has no effect unless multiple files are affected.
.Ar inc
must be an integer, and may be negative (producing progressively
earlier times).
.It Fl f
Attempt to force the update, even if the file permissions do not
currently permit it.
.It Fl h
If
.Ar file
is a symbolic link, access and/or modification time of the link is changed.
This option implies
.Fl c .
.It Fl i
Indirect: any
.Ar file
arguments are not files to be manipulated; rather, they name files
containing filenames, one per line, which are to be manipulated.  As a
special case, an argument of
.Sq \&\-
causes names to be read from the standard input.  (This special case
does not apply unless
.Fl i
is given.)
This is somewhat like using
.Xr xargs 1
with
.Nm touch ;
this option exists because (a) it is necessary to allow
.Fl d
to operate properly even on very long lists of files and (b)
.Xr xargs 1
considers space and tab, as well as newline, to delimit arguments.
.It Fl m
Change the modification time of the file.
The access time of the file is not changed unless the
.Fl a
flag is also specified.
.It Fl r
Use the access and modifications times from the specified file
instead of the current time of day.
.It Fl t
Change the access and modification times to the specified time.
The argument should be in the form
.Dq [[CC]YY]MMDDhhmm[.SS]
where each pair of letters represents the following:
.Pp
.Bl -tag -width Ds -compact -offset indent
.It Ar CC
The first two digits of the year (the century).
.It Ar YY
The second two digits of the year.
If
.Dq YY
is specified, but
.Dq CC
is not, a value for
.Dq YY
between 69 and 99 results in a
.Dq CC
value of 19.
Otherwise, a
.Dq CC
value of 20 is used.
.It Ar MM
The month of the year, from 1 to 12.
.It Ar DD
The day of the month, from 1 to 31.
.It Ar hh
The hour of the day, from 0 to 23.
.It Ar mm
The minute of the hour, from 0 to 59.
.It Ar SS
The second of the minute, from 0 to 61.
.El
.Pp
If the
.Dq CC
and
.Dq YY
letter pairs are not specified, the values default to the current
year.
If the
.Dq SS
letter pair is not specified, the value defaults to 0.
.El
.Pp
The
.Nm
utility exits 0 on success, and \*[Gt]0 if an error occurs.
.Sh SEE ALSO
.Xr utimes 2
.Sh COMPATIBILITY
The obsolescent form of
.Nm touch ,
where a time format is specified as the first argument, is supported.
When no
.Fl r
or
.Fl t
option is specified, there are at least two arguments, and the first
argument is a string of digits either eight or ten characters in length,
the first argument is interpreted as a time specification of the form
.Dq MMDDhhmm[YY] .
.Pp
The
.Dq MM ,
.Dq DD ,
.Dq hh
and
.Dq mm
letter pairs are treated as their counterparts specified to the
.Fl t
option.
If the
.Dq YY
letter pair is in the range 69 to 99, the year is set to 1969 to 1999,
otherwise, the year is set in the 21st century.
.Sh STANDARDS
The
.Nm
utility is expected to be a superset of the
.St -p1003.2
specification.  The
.Fl i
and
.Fl d
options are
.Nx
extensions.
.Sh HISTORY
A
.Nm
utility appeared in
.At v7 .
.Sh BUGS
A symbolic link can't be a reference file of access and/or modification
time.
.Pp
There is no way to get a newline into an indirect filename when using
.Fl i .
(Something analogous to
.Xr xargs 1 Ap s
.Fl 0
option would make sense here, but is of no real use in the anticipated
use case for
.Fl i ,
which involves using line-based text tools to process lists of
filenames.)
.Pp
Invalid arguments to
.Fl d
(eg, non-numbers) do not produce warning messages, instead silently
using whatever
.Xr atol 3
returns.
.Pp
NULs in
.Fl i
input will both terminate the pathname and cause the rest of the line
to be ignored.  The first part is not fixable unless and until we have
filesystem interfaces that do not depend on NUL-terminated strings; the
second part does not strike me as worth fixing.
.Pp
A partial line at the end of input (ie, if there is input but it does
not end with a newline) is ignored (with a warning message) when
reading filenames for
.Fl i .
Whether effectively supplying a newline is better is debatable.


usr.bin/touch/touch.c
/*	$NetBSD: touch.c,v 1.26 2006/03/18 11:15:00 dsl Exp $	*/

/*
 * Copyright (c) 1993
 *	The Regents of the University of California.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include <sys/cdefs.h>
#ifndef lint
__COPYRIGHT("@(#) Copyright (c) 1993\n\
	The Regents of the University of California.  All rights reserved.\n");
#endif /* not lint */

#ifndef lint
#if 0
static char sccsid[] = "@(#)touch.c	8.2 (Berkeley) 4/28/95";
#endif
__RCSID("$NetBSD: touch.c,v 1.26 2006/03/18 11:15:00 dsl Exp $");
#endif /* not lint */

#include <sys/types.h>
#include <sys/stat.h>
#include <sys/time.h>

#include <err.h>
#include <errno.h>
#include <fcntl.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <locale.h>
#include <time.h>
#include <tzfile.h>
#include <unistd.h>

int	main __P((int, char **));
int	rw __P((const char *, struct stat *, int));
void	stime_arg1 __P((char *, struct timeval *));
void	stime_arg2 __P((char *, int, struct timeval *));
void	stime_file __P((char *, struct timeval *));
void	usage __P((void));

static int (*change_file_times) __P((const char *, const struct timeval *));
static int (*get_file_status) __P((const char *, struct stat *));
static struct timeval tv[2];
static int aflag;
static int cflag;
static int fflag;
static int mflag;
static int rval;
static int timeset;
static long int inc;

static void touch_it(const char *fn)
{
	struct stat sb;
	int fd;

	/* See if the file exists. */
	if ((*get_file_status)(fn, &sb)) {
		if (cflag)
			return;
		/* Create the file. */
		fd = open(fn, O_WRONLY | O_CREAT, DEFFILEMODE);
		if (fd == -1 || fstat(fd, &sb) || close(fd)) {
			/* XXX this leaks the fd if fstat() fails;
			   is this possible enough we should care? */
			rval = 1;
			warn("%s", fn);
			return;
		}

		/* If using the current time, we're done. */
		if (!timeset)
			return;
	}
	if (!aflag)
		TIMESPEC_TO_TIMEVAL(&tv[0], &sb.st_atimespec);
	if (!mflag)
		TIMESPEC_TO_TIMEVAL(&tv[1], &sb.st_mtimespec);

	/* Try utimes(2). */
	if (!(*change_file_times)(fn, tv))
		return;

	/* If the user specified a time, nothing else we can do. */
	if (timeset) {
		rval = 1;
		warn("%s", fn);
		return;
	}

	/*
	 * System V and POSIX 1003.1 require that a NULL argument
	 * set the access/modification times to the current time.
	 * The permission checks are different, too, in that the
	 * ability to write the file is sufficient.  Take a shot.
	 */
	 if (!(*change_file_times)(fn, NULL))
		return;

	/* Try reading/writing. */
	if (!S_ISLNK(sb.st_mode) && rw(fn, &sb, fflag))
		rval = 1;
}

static void touch_file(const char *fn)
{
 touch_it(fn);
 tv[0].tv_sec += inc;
 tv[1].tv_sec += inc;
}

/* XXX Wouldn't normally use warn()/err(), but this program
   *already* depends on them, so little harm in this case. */
/* XXX Since this is for external consumption, we avoid lcs even
   though it would be the natural way to handle the control flow. */
static void touch_indirect(const char *fn)
{
 FILE *f;
 char *b;
 int a;
 int l;
 int c;

 f = strcmp(fn,"-") ? fopen(fn,"r") : stdin;
 if (! f)
  { warn("%s",fn);
    return;
  }
 b = 0;
 a = 0;
 l = 0;
 while /*<"read">*/ (1)
  { c = getc(f);
    if (a >= l)
     { b = realloc(b,a=l+16);
       if (! b) err(1,"realloc");
     }
    b[l++] = c; /* unnecessary but harmless if c==EOF */
    switch (c)
     { case EOF:
	  if (l > 1) warnx("%s: partial line ignored at end of input",fn);
	  /*break <"read">;*/goto break_read;
	  break;
       case '\n':
	  b[l-1] = '\0';
	  touch_file(b);
	  l = 0;
	  break;
     }
  }
break_read:;
 if (f != stdin) fclose(f);
 free(b);
 return;
}

int
main(argc, argv)
	int argc;
	char *argv[];
{
	int hflag, iflag, ch, len;
	char *p;

	setlocale(LC_ALL, "");

	aflag = cflag = fflag = hflag = mflag = timeset = 0;
	iflag = 0;
	inc = 0;
	if (gettimeofday(&tv[0], NULL))
		err(1, "gettimeofday");

	while ((ch = getopt(argc, argv, "acd:fhimr:t:")) != -1)
		switch(ch) {
		case 'a':
			aflag = 1;
			break;
		case 'c':
			cflag = 1;
			break;
		case 'd':
			inc = atol(optarg);
			break;
		case 'f':
			fflag = 1;
			break;
		case 'h':
			hflag = 1;
			break;
		case 'i':
			iflag = 1;
			break;
		case 'm':
			mflag = 1;
			break;
		case 'r':
			timeset = 1;
			stime_file(optarg, tv);
			break;
		case 't':
			timeset = 1;
			stime_arg1(optarg, tv);
			break;
		case '?':
		default:
			usage();
		}
	argc -= optind;
	argv += optind;

	/* Default is both -a and -m. */
	if (aflag == 0 && mflag == 0)
		aflag = mflag = 1;

	if (hflag) {
		cflag = 1;		/* Don't create new file */
		change_file_times = lutimes;
		get_file_status = lstat;
	} else {
		change_file_times = utimes;
		get_file_status = stat;
	}

	/*
	 * If no -r or -t flag, at least two operands, the first of which
	 * is an 8 or 10 digit number, use the obsolete time specification.
	 */
	if (!timeset && argc > 1) {
		(void)strtol(argv[0], &p, 10);
		len = p - argv[0];
		if (*p == '\0' && (len == 8 || len == 10)) {
			timeset = 1;
			stime_arg2(*argv++, len == 10, tv);
		}
	}

	/* Otherwise use the current time of day. */
	if (!timeset)
		tv[1] = tv[0];

	if (*argv == NULL)
		usage();

	for (rval = 0; *argv; ++argv) {
		if (iflag)
			touch_indirect(*argv);
		else
			touch_file(*argv);
	}
	exit(rval);
}

#define	ATOI2(s)	((s) += 2, ((s)[-2] - '0') * 10 + ((s)[-1] - '0'))

void
stime_arg1(arg, tvp)
	char *arg;
	struct timeval *tvp;
{
	struct tm *t;
	time_t tmptime;
	int yearset;
	char *p;
					/* Start with the current time. */
	tmptime = tvp[0].tv_sec;
	if ((t = localtime(&tmptime)) == NULL)
		err(1, "localtime");
					/* [[CC]YY]MMDDhhmm[.SS] */
	if ((p = strchr(arg, '.')) == NULL)
		t->tm_sec = 0;		/* Seconds defaults to 0. */
	else {
		if (strlen(p + 1) != 2)
			goto terr;
		*p++ = '\0';
		t->tm_sec = ATOI2(p);
	}
		
	yearset = 0;
	switch (strlen(arg)) {
	case 12:			/* CCYYMMDDhhmm */
		t->tm_year = ATOI2(arg) * 100 - TM_YEAR_BASE;
		yearset = 1;
		/* FALLTHROUGH */
	case 10:			/* YYMMDDhhmm */
		if (yearset) {
			t->tm_year += ATOI2(arg);
		} else {
			yearset = ATOI2(arg);
			if (yearset < 69)
				t->tm_year = yearset + 2000 - TM_YEAR_BASE;
			else
				t->tm_year = yearset + 1900 - TM_YEAR_BASE;
		}
		/* FALLTHROUGH */
	case 8:				/* MMDDhhmm */
		t->tm_mon = ATOI2(arg);
		--t->tm_mon;		/* Convert from 01-12 to 00-11 */
		/* FALLTHROUGH */
	case 6:
		t->tm_mday = ATOI2(arg);
		/* FALLTHROUGH */
	case 4:
		t->tm_hour = ATOI2(arg);
		/* FALLTHROUGH */
	case 2:
		t->tm_min = ATOI2(arg);
		break;
	default:
		goto terr;
	}

	t->tm_isdst = -1;		/* Figure out DST. */
	tvp[0].tv_sec = tvp[1].tv_sec = mktime(t);
	if (tvp[0].tv_sec == -1)
terr:		errx(1,
	"out of range or illegal time specification: [[CC]YY]MMDDhhmm[.SS]");

	tvp[0].tv_usec = tvp[1].tv_usec = 0;
}

void
stime_arg2(arg, year, tvp)
	char *arg;
	int year;
	struct timeval *tvp;
{
	struct tm *t;
	time_t tmptime;
					/* Start with the current time. */
	tmptime = tvp[0].tv_sec;
	if ((t = localtime(&tmptime)) == NULL)
		err(1, "localtime");

	t->tm_mon = ATOI2(arg);		/* MMDDhhmm[yy] */
	--t->tm_mon;			/* Convert from 01-12 to 00-11 */
	t->tm_mday = ATOI2(arg);
	t->tm_hour = ATOI2(arg);
	t->tm_min = ATOI2(arg);
	if (year) {
		year = ATOI2(arg);
		if (year < 69)
			t->tm_year = year + 2000 - TM_YEAR_BASE;
		else
			t->tm_year = year + 1900 - TM_YEAR_BASE;
	}
	t->tm_sec = 0;

	t->tm_isdst = -1;		/* Figure out DST. */
	tvp[0].tv_sec = tvp[1].tv_sec = mktime(t);
	if (tvp[0].tv_sec == -1)
		errx(1,
	"out of range or illegal time specification: MMDDhhmm[yy]");

	tvp[0].tv_usec = tvp[1].tv_usec = 0;
}

void
stime_file(fname, tvp)
	char *fname;
	struct timeval *tvp;
{
	struct stat sb;

	if (stat(fname, &sb))
		err(1, "%s", fname);
	TIMESPEC_TO_TIMEVAL(&tvp[0], &sb.st_atimespec);
	TIMESPEC_TO_TIMEVAL(&tvp[1], &sb.st_mtimespec);
}

int
rw(fname, sbp, force)
	const char *fname;
	struct stat *sbp;
	int force;
{
	int fd, needed_chmod, rval;
	u_char byte;

	/* Try regular files and directories. */
	if (!S_ISREG(sbp->st_mode) && !S_ISDIR(sbp->st_mode)) {
		warnx("%s: %s", fname, strerror(EFTYPE));
		return (1);
	}

	needed_chmod = rval = 0;
	if ((fd = open(fname, O_RDWR, 0)) == -1) {
		if (!force || chmod(fname, DEFFILEMODE))
			goto err;
		if ((fd = open(fname, O_RDWR, 0)) == -1)
			goto err;
		needed_chmod = 1;
	}

	if (sbp->st_size != 0) {
		if (read(fd, &byte, sizeof(byte)) != sizeof(byte))
			goto err;
		if (lseek(fd, (off_t)0, SEEK_SET) == -1)
			goto err;
		if (write(fd, &byte, sizeof(byte)) != sizeof(byte))
			goto err;
	} else {
		if (write(fd, &byte, sizeof(byte)) != sizeof(byte)) {
err:			rval = 1;
			warn("%s", fname);
		} else if (ftruncate(fd, (off_t)0)) {
			rval = 1;
			warn("%s: file modified", fname);
		}
	}

	if (fd >= 0 && close(fd) && rval != 1) {
		rval = 1;
		warn("%s", fname);
	}
	if (needed_chmod && chmod(fname, sbp->st_mode) && rval != 1) {
		rval = 1;
		warn("%s: permissions modified", fname);
	}
	return (rval);
}

__dead void
usage()
{
	(void)fprintf(stderr,
	    "usage: touch [-acfhim] [-r file] [-t time] [-d delta] file ...\n");
	exit(1);
}
